lib/curl_multibyte.c

```diff
@@ -88,12 +88,106 @@ char *curlx_convert_wchar_to_UTF8(const wchar_t *str_w)
 
 #if defined(USE_WIN32_LARGE_FILES) || defined(USE_WIN32_SMALL_FILES)
 
+
+/* Fix excessive paths (paths that exceed MAX_PATH length of 260).
+ *
+ * This is a helper function to fix paths that would exceed the MAX_PATH
+ * limitation check done by Windows APIs. It does so by normalizing the passed
+ * in filename or path 'in' to its full canonical path, and if that path is
+ * longer than MAX_PATH then setting 'out' to "\\?\" prefix + that full path.
+ *
+ * For example 'in' filename255chars in current directory C:\foo\bar is
+ * fixed as \\?\C:\foo\bar\filename255chars for 'out' which will tell Windows
+ * it's ok to access that filename even though the actual full path is longer
+ * than 255 chars.
+ *
+ * For non-Unicode builds this function may fail sometimes because only the
+ * Unicode versions of some Windows API functions can access paths longer than
+ * MAX_PATH, for example GetFullPathNameW which is used in this function. When
+ * the full path is then converted from Unicode to multibyte that fails if any
+ * directories in the path contain characters not in the current codepage.
+ */
+static bool fix_excessive_path(const TCHAR *in, TCHAR **out)
+{
+  size_t needed, written;
+  const wchar_t *in_w;
+  wchar_t *fbuf = NULL;
+
+#ifndef _UNICODE
+  wchar_t *ibuf = NULL;
+  char *obuf = NULL;
+#endif
+
+  *out = NULL;
+
+#ifndef _UNICODE
+  /* convert multibyte input to unicode */
+  needed = mbstowcs(NULL, in, 0);
+  if(needed == (size_t)-1 || needed >= (32767 - 4))
+    goto error;
+  ++needed; /* for NUL */
+  ibuf = malloc(needed * sizeof(wchar_t));
+  if(!ibuf)
+    goto error;
+  written = mbstowcs(ibuf, in, needed);
+  if(written == (size_t)-1 || written >= needed)
+    goto error;
+  in_w = ibuf;
+#else
+  in_w = in;
+#endif
+
+  /* get full unicode path of the unicode filename or path */
+  needed = (size_t)GetFullPathNameW(in_w, 0, NULL, NULL);
```

Comment:
Perhaps we should mention in the comment that `GetFullPathNameW` is also being used to translate path slashes such as `\.\` and `\..\` and to convert forward slashes to backslashes. 

Although the MS docs don't explicitly state it, I ran a simple test using:
```c++
#include <cstring>
#include <windows.h>
#include <iostream>

int main() {
  static const char a[] = "c:\\test\\..\\hello";
  char o[1000]{};
  GetFullPathNameA(a, sizeof(o), o, nullptr);
  std::cout << a << " -> " << o << std::endl;

  static const char b[] = "c:\\test\\../hello/./";
  char o2[1000]{};
  GetFullPathNameA(b, sizeof(o2), o2, nullptr);
  std::cout << b << " -> " << o2 << std::endl;
  return 0;
}
```
Output:
```bash
PS D:\test1> .\build\Debug\main.exe
c:\test\..\hello -> c:\hello
c:\test\../hello/./ -> c:\hello\
```

Reply:
> Perhaps we should mention in the comment that `GetFullPathNameW` is also being used to translate path slashes such as `\.\` and `\..\` and to convert forward slashes to backslashes.

Sure yes. That is normalization and I didn't explain exactly what it was but I changed the comment in the latest commit to address this:

~~~c
  /* GetFullPathNameW returns the normalized full path in unicode. It converts
     forward slashes to backslashes, processes .. to remove directory segments,
     etc. Unlike GetFullPathNameA it can process paths that exceed MAX_PATH. */
  needed = (size_t)GetFullPathNameW(in_w, 0, NULL, NULL);
~~~

Reply:
Looks great!